### PR TITLE
Fast Proposal Viewing

### LINKED ISSUE #10481: `governance:show` CLI command is slow
Copy/pasting notes from this [Slack thread](https://celo-org.slack.com/archives/C014SRL4URY/p1691476764159769):

> Does anyone know why the `governance:show` cli command is so slow nowadays?

> I run into the same, so I just checked this with `DEBUG=*`  and (if I read the logs correctly, I’m not very familiar with this), the cli is doing almost 1200 requests to answer.

> First guess would be it iterates over each potential proposal number (there have been ~120 proposals, so would match your number if there's around 10 txs per proposal). Or something with iterating over all validators. And potential retries?

```zsh
DEBUG=* ./bin/run governance:show --proposalID=121 --node https://forno.celo.org/
```

> Looks like there's a ton of requests to `0x41822d8a191fcfb1cfca5f7048818acd8ee933d3`, the GovernanceApproverMultiSigProxy.

### PR DESCRIPTION
**Pull request by aaronmgdr**
### Description

Speeds up several functions related to viewing a governance proposal by 

* running requests in parallel 
* avoiding some requests for data that was later discarded
* caching connection.chainId 
* adding `getContractMappingFromSourcify` result to the `addressMapping` Map which already stores similar data from CoreContracts. 

### Tested
```bash
$ yarn --cwd packages/cli run --silent celocli governance:show --proposalID 139
getGovernance: 103.049ms
Running Checks:
   ✔  139 is an existing proposal
All checks passed
newCheckBuilder: 89.329ms
getProposalMetaStage(139): 898.476ms
getApprovalStatus(139): 1.407s
getProposalState(139): 1.410s
getProposalRecord(139): 2.309s
getProposalRecord: 2.310s
proposalToJSON: 2.869s
getSupportWithConstitutionThreshold: 2.375s
humanReadableProposalSchedule: 424.337ms
run time: 8.176s
✨  Done in 18.98s.
```

### Related issues

- Fixes [#cli goverance:show slow](https://github.com/celo-org/celo-monorepo/issues/10481)

### Backwards compatibility

if the tests are good. 

### Documentation

n/a per only!

### CODE REVIEW

File: packages/sdk/explorer/src/block-explorer.ts [hunk 1]
```diff
@@ -315,8 +315,16 @@ export class BlockExplorer {
   getContractMappingFromSourcify = async (
     address: string
   ): Promise<ContractMapping | undefined> => {
+    const cached = this.addressMapping.get(address)
```
Comment (aaronmgdr) [inline packages/sdk/explorer/src/block-explorer.ts:318]: TODO would there be a reason why mappings from sourcify should not be cached in the way core contracts are?

File: packages/sdk/contractkit/src/wrappers/MultiSig.ts [hunk 2]
```diff
@@ -81,31 +83,59 @@ export class MultiSigWrapper extends BaseWrapper<MultiSig> {
...
+          const tx = await this.getTransaction(index, false)
+          if (tx.data === data && tx.destination === destination && tx.value.isEqualTo(value)) {
+            return index
+          }
+          return null
+        })
+    )
+    const txID = transactionIndexes.find((index) => index !== null)
+    if (!txID) {
+      return
```
Comment (nicolasbrugneaux) [inline packages/sdk/contractkit/src/wrappers/MultiSig.ts]: Aren't we overfetching now?

File: packages/sdk/explorer/src/block-explorer.ts (hunk 1 above)
Comment (nicolasbrugneaux) [reply packages/sdk/explorer/src/block-explorer.ts:318]: 🤷

File: packages/sdk/explorer/src/block-explorer.ts [hunk 3]
```diff
@@ -369,11 +377,14 @@ export class BlockExplorer {
...
-    }
+    const mappings = await Promise.all(
+      strategies.map(async (strategy) => {
+        const contractMapping = await strategy(address)
+        if (contractMapping && contractMapping.fnMapping.get(selector)) {
+          return contractMapping
+        }
+      })
+    )
+    return mappings.find((mapping) => mapping !== undefined)
```
Comment (nicolasbrugneaux) [inline packages/sdk/explorer/src/block-explorer.ts:388]: Same, this is maybe faster in some instance but it's over fetching

**Review by nicolasbrugneaux — APPROVED** [review]
Looks mostly good except 2 instances where I think it's maybe not _always_ faster to promise.all, because the for-loop can end as early as the first item, maybe cancellable promises here would help but that's a bigger refactor. Also it's a nitpick so feel free to discard

File: packages/sdk/contractkit/src/wrappers/MultiSig.ts (hunk 2 above)
Comment (aaronmgdr) [reply packages/sdk/contractkit/src/wrappers/MultiSig.ts]: true, would be better to add a way to just fetch confirmations for a transaction

**Comment by bowd** [comment]
@aaronmgdr this introduces a regression. As part of a proposal sometimes you want to update a proxy to a new implementation and call a method on that new implementation. This is done internally by keeping a mapping of updated implementations when they are found, but this requires a proposal to be parsed sequentially.

**Comment by bowd** [comment]
And the annoying thing is that because `celocli` has  the @celo/governance dependency specified as `^5.0.4`, even if I install an older (3.0.1) version of the cli it still will pull `5.0.6` as the governance package and have parallelization. We should probably pin internal dependencies :(

**Pull request merged**